engine/schema/src/com/cloud/upgrade/DatabaseVersionHierarchy.java

```diff
@@ -0,0 +1,185 @@
+
+        if (toVersion == null) {
+            // all the available versions greater than or equal to fromVersion
+            predicate = node -> node.version.compareTo(fromVersion) > -1;
+        } else {
+            // all the available versions greater than or equal to fromVersion AND less than toVersion
+            predicate = node -> node.version.compareTo(fromVersion) > -1 && node.version.compareTo(toVersion) < 0;
+        }
+
+        // get upgrade path from version forward (include version itself in the path)
+        return hierarchy
+                    .stream()
+                    .filter(predicate)
+                    .filter(distinct(node -> node.upgrader.getUpgradedVersion()))
+                    .map(node -> node.upgrader)
+                    .toArray(DbUpgrade[]::new);
+    }
+
+    /**
+     * Find the most recent <code>CloudStackVersion</code> immediately before <code>fromVersion</code>
+     *
+     * @param fromVersion The version to look up its immediate previous available version
+     *
+     * @return The <code>CloudStackVersion</code> or null
+     *
+     * @since 4.8.2.0 (refactored in 4.11.1.0)
+     */
+    private CloudStackVersion getRecentVersion(final CloudStackVersion fromVersion) {
+        if (fromVersion == null) {
+            return null;
+        }
+
+        // find the most recent version immediately before fromVersion
+        return toList()
+                .reverse()
+                .stream()
+                .filter(version -> fromVersion.compareTo(version) < 0)
+                .findFirst()
+                .orElse(null);
+    }
+
+    /**
+     * Generate immutable list of available <code>CloudstackVersion</code> in the hierarchy
+     *
+     * @return list of available versions
+     */
+    public ImmutableList<CloudStackVersion> toList() {
+        List<CloudStackVersion> versions = hierarchy
+                                                .stream()
+                                                .map(node -> node.version)
+                                                .collect(Collectors.toList());
+
+        return ImmutableList.copyOf(versions);
+    }
+
+    /**
+     * Find the distinct <code>VersionNode</code> based on the provided <code>getUpgradedVersion()</code>
+     */
+    private Predicate<VersionNode> distinct(Function<VersionNode, String> keyExtractor) {
+        Map<String, Boolean> seen = new ConcurrentHashMap<>();
+        return t -> seen.putIfAbsent(keyExtractor.apply(t), Boolean.TRUE) == null;
+    }
+
+    private static class VersionNode {
+        final CloudStackVersion version;
+        final DbUpgrade upgrader;
+
+        private VersionNode(final CloudStackVersion version, final DbUpgrade upgrader) {
+            this.version = version;
+            this.upgrader = upgrader;
+        }
+    }
+
+    public static final class DatabaseVersionHierarchyBuilder {
+        private final List<VersionNode> _hierarchyBuilder = new LinkedList<>();
```

Comment:
You do not need to use these "_" anymore.

Is there a reason for you to use `LinkedList` here?